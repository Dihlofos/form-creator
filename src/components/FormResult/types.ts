export type FormLocalState = Record<string, string | boolean>;
