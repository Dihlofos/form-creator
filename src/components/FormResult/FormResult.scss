.form-result {
  &__field {
    margin: 15px 0;
  }

  &__group {
    padding: 15px;
    background-color: var(--tooltipBg);
  }

  &__buttons {
    display: flex;
    flex-wrap: wrap;
    gap: 20px;
  }
}
