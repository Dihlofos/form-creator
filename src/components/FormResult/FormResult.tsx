import React, { FC, useCallback, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  doCustomAction,
  getFormData,
  getFormLocalState,
  setFormLocalState,
} from 'store';
import { Button } from 'ui';
import FormField from '../FormField/FormField';

import './FormResult.scss';

const FormResult: FC = () => {
  const form = useSelector(getFormData);
  const formState = useSelector(getFormLocalState);
  const dispatch = useDispatch();
  const { title, items, buttons, action: formAction, method } = form;

  const handleButtonClick = useCallback(
    (action: string | undefined) => () => {
      if (action) {
        dispatch(doCustomAction(action));
      }
    },
    [dispatch]
  );

  const handleChange = useCallback(
    (e) => {
      const { value, name, checked } = e.target;

      if (!value) {
        dispatch(setFormLocalState({ ...formState, [name]: checked }));
        return;
      }

      dispatch(setFormLocalState({ ...formState, [name]: value }));
    },
    [dispatch, formState]
  );

  const renderFields = useMemo(
    () => (
      <form
        className="form-result__form"
        action={formAction && formAction}
        method={method && method}
      >
        {items.map((field) => (
          <div className="form-result__field" key={field.id}>
            <FormField
              field={field}
              onChange={handleChange}
              value={formState[field.name]}
              checked={formState[field.name]}
            />
          </div>
        ))}
      </form>
    ),
    [items, handleChange, formAction, method, formState]
  );

  const renderButtons = useMemo(
    () => (
      <div className="form-result__buttons">
        {buttons &&
          buttons.map((button) => (
            <div className="form-result__field" key={button.id}>
              <Button onClick={handleButtonClick(button.action)}>
                {button.label}
              </Button>
            </div>
          ))}
      </div>
    ),
    [handleButtonClick, buttons]
  );

  return (
    <div className="form-result">
      <h2 className="form-result__title">{title}</h2>
      {renderFields}
      {buttons && renderButtons}
    </div>
  );
};

export default FormResult;
