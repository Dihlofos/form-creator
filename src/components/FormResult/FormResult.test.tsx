import React from 'react';
import { create } from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import InitialState from '../../jest/InitialState';
import FormResult from './FormResult';

const mockStore = configureMockStore();

describe('FormResult test', () => {
  it('snapshot test', () => {
    const store = mockStore(InitialState);

    const tree = create(
      <Provider store={store}>
        <FormResult />
      </Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
