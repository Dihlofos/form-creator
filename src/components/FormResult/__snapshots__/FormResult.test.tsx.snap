// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormResult test snapshot test 1`] = `
<div
  className="form-result"
>
  <h2
    className="form-result__title"
  >
    Тестовая форма
  </h2>
  <form
    action=""
    className="form-result__form"
    method=""
  >
    <div
      className="form-result__field"
    >
      <div
        className="wrapper-input-component"
      >
        <div
          className="ant-row ant-form-item label-input-component"
          style={Object {}}
        >
          <div
            className="ant-col ant-form-item-label"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="name"
              title="Имя"
            >
              Имя
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              />
            </div>
          </div>
        </div>
        <input
          className="ant-input input-component"
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Введите имя"
          type="text"
          value="undefined"
        />
      </div>
    </div>
    <div
      className="form-result__field"
    >
      <div
        className="wrapper-input-component"
      >
        <div
          className="ant-row ant-form-item label-input-component"
          style={Object {}}
        >
          <div
            className="ant-col ant-form-item-label"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="number"
              title="Числа"
            >
              Числа
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              />
            </div>
          </div>
        </div>
        <input
          className="ant-input input-component"
          id="number"
          name="number"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Введите число"
          type="number"
          value="undefined"
        />
      </div>
    </div>
    <div
      className="form-result__field"
    >
      <div
        className="ant-row ant-form-item label-input-component"
        style={Object {}}
      >
        <div
          className="ant-col ant-form-item-label"
          style={Object {}}
        >
          <label
            className=""
            htmlFor="textarea"
            title="Опишите нечто"
          >
            Опишите нечто
          </label>
        </div>
        <div
          className="ant-col ant-form-item-control"
          style={Object {}}
        >
          <div
            className="ant-form-item-control-input"
          >
            <div
              className="ant-form-item-control-input-content"
            />
          </div>
        </div>
      </div>
      <textarea
        className="ant-input textarea"
        id="textarea"
        label="Опишите нечто"
        name="textarea"
        onChange={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onKeyDown={[Function]}
        placeholder="Введите многострочный текст"
        style={Object {}}
        value="undefined"
      />
    </div>
    <div
      className="form-result__field"
    >
      <label
        className="ant-checkbox-wrapper"
      >
        <span
          className="ant-checkbox"
          style={Object {}}
        >
          <input
            checked={false}
            className="ant-checkbox-input"
            id="checkbox"
            name="checkbox"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyPress={[Function]}
            onKeyUp={[Function]}
            type="checkbox"
          />
          <span
            className="ant-checkbox-inner"
          />
        </span>
        <span>
          Чекбокс
        </span>
      </label>
    </div>
    <div
      className="form-result__field"
    >
      <div
        className="wrapper-input-component"
      >
        <div
          className="ant-row ant-form-item label-input-component"
          style={Object {}}
        >
          <div
            className="ant-col ant-form-item-label"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="date"
              title="Дата"
            >
              Дата
            </label>
          </div>
          <div
            className="ant-col ant-form-item-control"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              />
            </div>
          </div>
        </div>
        <input
          className="ant-input input-component"
          id="date"
          name="date"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Введите дату"
          type="date"
          value="undefined"
        />
      </div>
    </div>
    <div
      className="form-result__field"
    >
      <h4>
        Как вам в IT?
      </h4>
      <div
        className="ant-radio-group ant-radio-group-outline"
        id="howit"
      >
        <label
          className="ant-radio-wrapper"
        >
          <span
            className="ant-radio"
            style={Object {}}
          >
            <input
              checked={false}
              className="ant-radio-input"
              name="howit"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="great"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            Отлично, нравится
          </span>
        </label>
        <label
          className="ant-radio-wrapper"
        >
          <span
            className="ant-radio"
            style={Object {}}
          >
            <input
              checked={false}
              className="ant-radio-input"
              name="howit"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="acceptable"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            Приемлемо
          </span>
        </label>
        <label
          className="ant-radio-wrapper"
        >
          <span
            className="ant-radio"
            style={Object {}}
          >
            <input
              checked={false}
              className="ant-radio-input"
              name="howit"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              type="radio"
              value="burned"
            />
            <span
              className="ant-radio-inner"
            />
          </span>
          <span>
            Выгорел
          </span>
        </label>
      </div>
    </div>
  </form>
  <div
    className="form-result__buttons"
  >
    <div
      className="form-result__field"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <span>
          Отправить
        </span>
      </button>
    </div>
    <div
      className="form-result__field"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <span>
          Сбросить
        </span>
      </button>
    </div>
    <div
      className="form-result__field"
    >
      <button
        className="ant-btn"
        onClick={[Function]}
        type="button"
      >
        <span>
          Редактировать
        </span>
      </button>
    </div>
  </div>
</div>
`;
